Partially revert "libnet: remove struct endpointCnt"

akerouanton opened the pull request:
Related to:

- https://github.com/moby/moby/pull/49736
- https://github.com/moby/moby/pull/49773
- https://github.com/moby/moby/pull/49769

**- What I did**

This change was causing a panic when downgrading from RC1 to an older Engine release.

We still want to keep the new endpoint / network caching behavior introduced in that commit as it might fix some instances of the 'has active endpoints' bug, and it's needed for the error message improvement made in 241d685.

So, instead of doing a full revert of commit 51d7f95c4b364d564d4653a7915112120385e7cd, restore the endpointCnt struct and persist it, but do not use it to determine if a network is in use.

Moreover, commit 78be7ebad added a Context arg to `NewNetwork`, so use it in the code re-introduced.

For reference, here's the observed stack trace:

```
goroutine 1 [running, locked to thread]:
github.com/docker/docker/libnetwork.(*endpointCnt).EndpointCnt(0x4000922cd0?)
	/root/build-deb/engine/libnetwork/endpoint_cnt.go:102 +0x24
github.com/docker/docker/libnetwork.(*Network).delete(0x400049f580?, 0x0, 0x0)
	/root/build-deb/engine/libnetwork/network.go:1048 +0x248
github.com/docker/docker/libnetwork.(*Network).Delete(0x40008aae00, {0x0, 0x0, 0x8?})
	/root/build-deb/engine/libnetwork/network.go:1012 +0x78
github.com/docker/docker/daemon.configureNetworking(0x40009540e0, 0x400047a608)
	/root/build-deb/engine/daemon/daemon_unix.go:884 +0x128
github.com/docker/docker/daemon.(*Daemon).initNetworkController(0x4000455d48, 0x400047a608, 0x40006895c0)
	/root/build-deb/engine/daemon/daemon_unix.go:858 +0x128
github.com/docker/docker/daemon.(*Daemon).restore(0x4000455d48, 0x400047a608)
 	/root/build-deb/engine/daemon/daemon.go:524 +0x518
```

**- How to verify it**

To observe the issue:

- Start the Engine from v28.1.0-rc.1
- Then, start the Engine from v28.0.4 and observe the panic

And to test the failure (you need to delete `/var/lib/docker/network/files/local-kv.db` first):

- Start the Engine from this branch
- Then, start the Engine from v28.0.4 -- it should not panic this time

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
- Make sure older Engine versions won't panic when downgrading from v28.1.
```

akerouanton committed:
Partially revert "libnet: remove struct endpointCnt"

This change was causing a panic when downgrading from RC1 to an older
Engine release.

We still want to keep the new endpoint / network caching behavior
introduced in that commit as it might fix some instances of the 'has
active endpoints' bug, and it's needed for the error message improvement
made in 241d685.

So, instead of doing a full revert of commit 51d7f95c4b364d564d4653a7915112120385e7cd,
restore the endpointCnt struct and persist it, but do not use it to
determine if a network is in use.

For reference, here's the observed stack trace:

```
goroutine 1 [running, locked to thread]:
github.com/docker/docker/libnetwork.(*endpointCnt).EndpointCnt(0x4000922cd0?)
	/root/build-deb/engine/libnetwork/endpoint_cnt.go:102 +0x24
github.com/docker/docker/libnetwork.(*Network).delete(0x400049f580?, 0x0, 0x0)
	/root/build-deb/engine/libnetwork/network.go:1048 +0x248
github.com/docker/docker/libnetwork.(*Network).Delete(0x40008aae00, {0x0, 0x0, 0x8?})
	/root/build-deb/engine/libnetwork/network.go:1012 +0x78
github.com/docker/docker/daemon.configureNetworking(0x40009540e0, 0x400047a608)
	/root/build-deb/engine/daemon/daemon_unix.go:884 +0x128
github.com/docker/docker/daemon.(*Daemon).initNetworkController(0x4000455d48, 0x400047a608, 0x40006895c0)
	/root/build-deb/engine/daemon/daemon_unix.go:858 +0x128
github.com/docker/docker/daemon.(*Daemon).restore(0x4000455d48, 0x400047a608)
 	/root/build-deb/engine/daemon/daemon.go:524 +0x518
```

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

vvoland commented:
Could we split this into 2 commits?

1. Revert of https://github.com/moby/moby/commit/51d7f95c4b364d564d4653a7915112120385e7cd (minus the context plumbing)
2. The delta between the revert and the new code

akerouanton committed:
libnetwork: NewNetwork: use ctx everywhere

Commit 78be7ebab added a Context arg to `NewNetwork`, and previous
commit reverted a change made before that commit. So use the new `ctx`
parameter in the code re-introduced.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton commented:
I'm closing this one in favor of https://github.com/moby/moby/pull/49812.
